<a href="https://colab.research.google.com/github/heisenberg304/Housing-Prices-prediction/blob/develop/Modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Modeling

│

├── Dummy Regressor


├── Linear Regression


├── Ridge Regression


├── Lasso Regression


├── Decision Tree Regressor


├── Random Forest Regressor


├── Gradient Boosting


└── XGBoost

#LOAD DATA

In [89]:
%%writefile data_loading.py
import pandas as pd

#cargar datos
# data train no filled
train_not_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_not_filled_encoding.csv")

# data train filled
train_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_train_filled_encoding.csv")

# data test no filled
test_not_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_test_not_filled_encoding.csv")

# data test filled
test_filled = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/HP_test_filled_encoding.csv")
y_test = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/y_test.csv")
test = pd.read_csv("/content/drive/MyDrive/house_prices_advanced/House_Prices_test.csv")
#(y_test["Id"] == test["Id"]).all()    True


#split data filled
y_train = train_filled["SalePrice"]
x_train = train_filled.drop("SalePrice", axis=1)
y_test = y_test.drop("Id", axis=1)
x_test = test_filled

#ordenar estos features
x_test = x_test[x_train.columns]


#split data not filled
y_train_nf = train_not_filled["SalePrice"]
x_train_nf = train_not_filled.drop("SalePrice", axis=1)
y_test_nf = y_test
x_test_nf = test_not_filled

#ordenar estos features
x_test_nf = x_test_nf[x_train_nf.columns]

#

Writing data_loading.py


In [90]:
%%writefile modeling_def.py
#from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score

#entrenar modelo
def entrenar_model(model, x, y):
  return model.fit(x,y)

#predicciones del modelo
def predict_model(model, x):
  return model.predict(x)

#metricas del modelo
def evaluate_model(y_pred, y_true):
  MSE = mean_squared_error(y_pred, y_true)
  RMSE = root_mean_squared_error(y_pred, y_true)
  MAE = mean_absolute_error(y_pred, y_true)
  R2 = r2_score(y_pred, y_true)
  return MSE, RMSE, MAE, R2

#

Writing modeling_def.py


#Baseline
   - Dummy

In [97]:
#Dummy Regressor
from data_loading import x_train, y_train, x_test, y_test
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = DummyRegressor()


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Root Mean Squared Error: 16603.77832552691
 Mean Absolute Error: 12084.41645179763
 R2 Score: -3.2547193775269945e+29 
Mean Squared Error: 275685454.6832372


#Linear Models
   - Linear
   - Ridge
   - Lasso

In [98]:
#Linear Regression
from sklearn.linear_model import LinearRegression
from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = LinearRegression()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Root Mean Squared Error: 574883.4453389066
 Mean Absolute Error: 566946.2718963341
 R2 Score: -30.74370422762672 
Mean Squared Error: 330490975724.73157


In [ ]:
#Ridge Regression
from sklearn.linear_model import Ridge
from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = Ridge()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

In [ ]:
#Lasso Regression
from sklearn.linear_model import Lasso
from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = Lasso()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

#Tree Models
   - Decision Tree Regressor
   - Random Forest Regressor

In [ ]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor
from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = DecisionTreeRegressor()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

In [ ]:
#Random Forest Regressor
from sklearn.ensemble import RandomForestRegressor
from data_loading import x_train, y_train, x_test, y_test
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = RandomForestRegressor()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Mean Squared Error: 4838813584.039933
 Root Mean Squared Error: 69561.58123590876
 Mean Absolute Error: 53178.7121148835
 R2 Score: 0.0784301118303089


#Boosting Models
   - Gradient Boosting Regressor
   - XGBoost Regressor

In [96]:
#Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor
from data_loading import x_train, y_train, x_test, y_test, y_train_nf, x_train_nf, y_test_nf, x_test_nf
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = GradientBoostingRegressor()
x_test = x_test[x_train.columns]


#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

#

 Root Mean Squared Error: 72989.81385674416
 Mean Absolute Error: 54961.249867338076
 R2 Score: 0.0839142099098249 
Mean Squared Error: 5327512926.842162


In [95]:
from xgboost import XGBRegressor
import data_loading
import importlib
importlib.reload(data_loading)
from data_loading import x_train, y_train, x_test, y_test, y_train_nf, x_train_nf, y_test_nf, x_test_nf
from sklearn.metrics import mean_squared_error, root_mean_squared_error, mean_absolute_error, r2_score
from modeling_def import entrenar_model, predict_model, evaluate_model

#modelo
model = XGBRegressor()
x_test = x_test[x_train.columns]

#entrenar modelo
modelo_entrenado = entrenar_model(model, x_train, y_train)

#predicciones del modelo
predict = predict_model(modelo_entrenado, x_test)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict, y_test)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

###################################################################################################################################################

#entrenar modelo
modelo_entrenado_nf = entrenar_model(model, x_train_nf, y_train_nf)

#predicciones del modelo
predict_nf = predict_model(modelo_entrenado_nf, x_test_nf)

#metricas del modelo
MSE, RMSE, MAE, R2 = evaluate_model(predict_nf, y_test_nf)

print(f' Root Mean Squared Error: {RMSE}\n Mean Absolute Error: {MAE}\n R2 Score: {R2} \nMean Squared Error: {MSE}')

 Mean Squared Error: 5553550848.0
 Root Mean Squared Error: 74522.1484375
 Mean Absolute Error: 55351.1015625
 R2 Score: 0.07572102546691895
 Root Mean Squared Error: 73868.5390625
 Mean Absolute Error: 54838.48828125
 R2 Score: 0.08003711700439453 
Mean Squared Error: 5456560640.0


#Tabla comparativa
---

| Modelo | RMSE ↓ | MAE ↓ | R² ↑ | MSE |
|---------|--------:|-------:|------:|:------:|
| Dummy |16603.778 |12084.416 |-3.255 |275685454.683 |
| Linear |574883.445 |566946.272 |-30.7437 |330490975724.73 |
| Ridge | | | | |
| Lasso | | | | |
| Decision Tree | | | | |
| Random Forest | | | | |
| Gradient Boosting | | | | |
| XGBoost | | | | |
| XGBoost_Not_Filled | | | | |
---

